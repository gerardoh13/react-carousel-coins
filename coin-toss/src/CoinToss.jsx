import React, { useState } from "react";
import Coin from "./Coin";

function CoinToss(props) {
  const [coin, setCoin] = useState(props.coins[0]);
  const [headsCount, setHeadsCount] = useState(0);
  const [tailsCount, setTailsCount] = useState(0);
  const [flipping, setFlipping] = useState(false);

  const animateFlip = () => {
    setFlipping(true);
    setTimeout(() => {
      setFlipping(false);
    }, 400);
  };

  const handleClick = () => {
    const randIdx = Math.floor(Math.random() * 2);
    const newCoin = props.coins[randIdx];
    setCoin(newCoin);
    if (newCoin.side === "heads") {
      setHeadsCount(headsCount + 1);
    } else {
      setTailsCount(tailsCount + 1);
    }
    if (!props.testing) {
      animateFlip();
    }
  };
  let firstFlip = (headsCount + tailsCount) === 0 ? true : false
  return (
    <>
      <h1 className="mb-3">Let's flip a coin!</h1>
      <Coin side={coin.side} imgSrc={coin.imgSrc} flip={flipping} first={firstFlip}/>
      <button
        className="btn btn-secondary my-3"
        onClick={handleClick}
        disabled={flipping}
      >
        FLIP MEEEE
      </button>
      <p>
        Out of {headsCount + tailsCount} flips, there have been {headsCount}{" "}
        heads and {tailsCount} tails.
      </p>
    </>
  );
}

CoinToss.defaultProps = {
  coins: [
    {
      side: "heads",
      imgSrc:
        "https://fortunetellingplus.com/assets/img/coins/Half-Dollar-2016-head.png?quality=50",
    },
    {
      side: "tails",
      imgSrc:
        "https://fortunetellingplus.com/assets/img/coins/Half-Dollar-2016-tail.png?quality=50",
    },
  ],
  testing: false,
};

export default CoinToss;
