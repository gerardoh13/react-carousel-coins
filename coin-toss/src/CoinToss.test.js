import React from "react";
import { render, fireEvent } from "@testing-library/react";
import CoinToss from "./CoinToss";

beforeEach(function() {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

it("renders without crashing", function() {
  render(<CoinToss testing/>);
});

it("matches snapshot", function() {
  const { asFragment } = render(<CoinToss testing/>);
  expect(asFragment()).toMatchSnapshot();
});

it("doesn't show a coin on page load", function() {
  const { queryByTestId } = render(<CoinToss testing/>);

  expect(queryByTestId("coin")).toHaveClass("hidden");
});

it("counts correctly when heads appears", function() {
  const { getByText, queryByAltText } = render(<CoinToss testing/>);

  const button = getByText("FLIP MEEEE");
  fireEvent.click(button);

  expect(queryByAltText("heads")).toBeInTheDocument();
  expect(queryByAltText("tails")).not.toBeInTheDocument();
  expect(
    getByText("Out of 1 flips, there have been 1 heads and 0 tails.")
  ).toBeInTheDocument();
});

it("counts correctly when tails appears", function() {
  const { getByText, queryByAltText } = render(<CoinToss testing/>);

  const button = getByText("FLIP MEEEE");
  fireEvent.click(button);
  fireEvent.click(button);

  expect(queryByAltText("tails")).toBeInTheDocument();
  expect(queryByAltText("heads")).not.toBeInTheDocument();
  expect(
    getByText("Out of 2 flips, there have been 1 heads and 1 tails.")
  ).toBeInTheDocument();
});

afterEach(function() {
  Math.random.mockRestore();
});
