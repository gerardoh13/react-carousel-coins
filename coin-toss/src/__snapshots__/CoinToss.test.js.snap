// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1
    class="mb-3"
  >
    Let's flip a coin!
  </h1>
  <div
    class="Coin"
  >
    <img
      alt="heads"
      class=" hidden"
      data-testid="coin"
      src="https://fortunetellingplus.com/assets/img/coins/Half-Dollar-2016-head.png?quality=50"
    />
  </div>
  <button
    class="btn btn-secondary my-3"
  >
    FLIP MEEEE
  </button>
  <p>
    Out of 0 flips, there have been 0 heads and 0 tails.
  </p>
</DocumentFragment>
`;
